Make LinearLSQFitter work properly with fixed parameters & allow fitting Shift with LinearLSQFitter & LevMarLSQFitter

jehturner opened the pull request:
This is a prospective fix for #6103 & #6135. I'll add a couple of tests next week but don't want to spring this on you closer to 2.0 than necessary 🙂. I found this harder than it looks, with complications like model sets with different axes, different `col_fit_deriv` conventions and input dimensionality to understand when doing the linear algebra, as well as and how some of the modeling internals are supposed to work and their existing limitations, but I'm pretty confident that it's a substantial improvement on the previous behaviour and it has been passing all the tests.

Does anyone know why the print statement in the example under *Fitters support constrained fitting* [here](https://github.com/astropy/astropy/blob/878184fb3f987f325b44687f9eda27f577330c92/docs/modeling/fitting.rst) has `# doctest: +SKIP`? Without that the doctest would verify that the results have changed as expected. Thanks!

cc: @nden, @pllim 
cc: @embray, as I believe he wrote some of the relevant infrastructure, particularly model sets, which I'm not sure I quite understand the intended behaviour of 100%.

jehturner commented:
Just FYI, I'll be heading out on a family trip in a week's time, until mid-late July (though I'll probably be scanning my email from time to time). Thanks!

jehturner commented:
Argh, what a mess. I was just trying to rebase this after something changed in master. I don't think the failures are real. I might add another test tomorrow anyway.

bsipocz commented:
@jehturner - Could you please rebase to be on top of master rather than pull in master?

jehturner commented:
@bsipocz: Yes, sorry, that's what I was trying to do but it ended up with merge commits due to conflicts both times (my workflow is probably a bit crap). I'm just about to add a couple more tests, then it should be done. Can you think of any quick way to clean this up? I think maybe I'd have to reset back to 0199660, rebase from master, re-do the last couple of commits and `push -f`?? Not sure what that does to the PR...

jehturner commented:
@bsipocz: That seemed to work fine; the PR history above is a bit of a mess (referring to a few commits that no longer exist), but I believe the git commit history is now clean and correct.

I have added a couple more tests and think this is finished, subject to any comments.

pllim commented on astropy/modeling/functional_models.py:
```diff
@@ -506,7 +506,7 @@ def _parameter_units_for_data_units(self, inputs_unit, outputs_unit):
 
 
-class Shift(Model):
+class Shift(Fittable1DModel):
```
Just want to note #4721 here since that one didn't appear to solve the problem.

jehturner replied:
Thanks, I hadn't spotted that. Although that PR was merged, there no longer seems to be any trace of it in the commit history, so presumably it was undone soon afterwards?

jehturner replied:
I should probably have added a `fit_deriv` for Scale while I was at it, otherwise we may get questions about why one works but not the other (there are probably other examples but those two go together).

pllim replied:
That PR only set `fittable = True` without changing inheritance.

jehturner replied:
Sorry, I misinterpreted your line comment. I'll add `Scale` as well.

jehturner commented:
I've made `Scale` fittable as well, as per #4721, pointed out by @pllim, added a couple more tests and rebased again. This should be ready for further review.

jehturner commented:
It's currently raining merge conflicts (mostly trivial ones in CHANGES.rst), so I'll wait for you to tell me you're ready to merge this before rebasing again (if you want), or you're welcome to do it if I'm travelling.

nden commented:
@jehturner This looks good and fixes a few bugs. Can you rebase to fix the conflicts and squash the commits? I think it's good to go in this release. Thanks for the work!

Out of curiosity, can you think of other models that may need `implicit_part`?

bsipocz commented:
@jehturner - If you do the interactive rebase, it offers you to squash and also to rewrite the commit message as well. I think an overall summary would be a good middle ground.

nden commented:
@jehturner Re: To me `implicit_part` did not sound obvious at all. I can't think of a better name. On the positive side it made me read the explanation carefully in order to understand what it is. I suggest to keep it this way if there are no better suggestions.

jehturner commented:
Is `implicit_term` any better?

nden commented:
`sum_of_implicts_terms` sounds most clear to me. I also considered whether it should be private and decided against it for the same reasons.

nden commented:
@jehturner I realize now you may not know that the deadline for 2.0 is today. Can you finish this today?

jehturner commented:
OK, sorry, I was in a meeting. I knew there was a deadline around now but thought it was mainly for new features. I'll do it in a minute then.

jehturner commented:
I'm just having trouble figuring out how to add a multi-line commit message in an interactive rebase and consequently getting in a tangle.

nden commented:
When I do `git rebase -i astropy/master` an editor window is opened and I can choose which commits to skip. Once this editor window is closed, another one opens with the commit messages. I can write there a multiline commit message or choose which previous commit messages to comment out.
Isn't this what you are seeing?

jehturner commented:
Thanks. Yes, that's what I just realized from some reading, but I hadn't got as far as the second editor window because the first one failed after I tried to add a multi-line commit message...

nden commented:
Yah, in the first window just mark with `s` the commit you want to skip. Write commit messages in the second window.

jehturner commented:
OK, I never got the second editor window but it's easy enough to change the last commit message. I'm just re-checking that it looks as expected before I push it.

jehturner commented:
Hmmm, the commit log got a bit messed up but it looks like that's the only problem. Just a minute...

nden commented:
LGTM Thanks @jehturner !
